### PR TITLE
Dialogflow Webhook Example in Python

### PR DESCRIPTION
**Pull request by carolinahdez**
Google’s Contact Center AI (CCAI) - Webhook design example
bug 152346888

### CODE REVIEW

File: examples/dialogflow-webhook-example/main.py [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+      #               {'type': 'deposit', 'amount': 90}
+      #            ],
+      #            'balance': 110
+      #         }
+      #     }
+      # }
+      user_dict = user.to_dict()
+      transactions = user_dict['accounts']['checking']['transactions']
+      message = 'From webhook: Here are all the transactions that I found:\n'
+      checking_transactions = ''
```
Comment (jaketf) [inline examples/dialogflow-webhook-example/main.py]: L300-319 is redundant, verbose, memory inefficient (allocating a list of two elements in every iteration just to put a comma between the latest item and the rest of the items aggregated in the string), slow (string.format is slower than f-string) and leads to some quite confusing variable naming / reuse. Instead this should be DRY-d up / greatly simplified using list concatenation with `+` operator and list comprehension. This means we we only allocate 1 list. This is much more readable (and faster)
```python3
      all_transactions = ','.join([
          '{} {}'.format(transaction['type'], transaction['amount'])
          for transaction in user_dict['accounts']['checking']['transactions'] +
          user_dict['accounts']['saving']['transactions']
      ])
      return json.dumps({
          'fulfillmentText':
              ' '.join([
                  message, all_transactions, 'What else can I do for you?'])})
```
and the following benchmark shows it's significantly faster :)
```
old:  9.91780741896946
new:  6.51355233398499
```

```python3 
import timeit
user_dict = {
    'accounts': {
        'checking': {
            'transactions': [{
                'type': 'foo',
                'amount': i
            } for i in range(10)]
        },
        'saving': {
            'transactions': [{
                'type': 'bar',
                'amount': i 
            } for i in range(10)]
        }
    }
}

def old():
    checking_transactions = ''
    saving_transactions = ''
    transactions = user_dict['accounts']['checking']['transactions']
    for transaction in transactions:
      checking_transactions = ','.join([
          '{} {}'.format(transaction['type'], transaction['amount']),
          checking_transactions
      ])
    transactions = user_dict['accounts']['saving']['transactions']
    for transaction in transactions:
      saving_transactions = ','.join([
          '{} {}'.format(transaction['type'], transaction['amount']),
          saving_transactions
      ])

def new():
    all_transactions = ','.join([
        f"{transaction['type']} {transaction['amount']}"
        for transaction in user_dict['accounts']['checking']['transactions'] +
        user_dict['accounts']['saving']['transactions']
    ])
print('old: ', timeit.timeit(old))
print('new: ', timeit.timeit(new))
```

File: examples/dialogflow-webhook-example/main.py [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+    # Get the user_id from the context in the request.
+    user_id = self._get_user_id(flattener)
+
+    # Get the firestore client and extract the user data base on user_id
+    # provided.
+    db = get_firestore_client()
+    client_ref = db.collection(u'users').where(u'user_id', u'==', user_id)
+    users = client_ref.stream()
+    for user in users:
+      # Example:
```
Comment (jaketf) [inline examples/dialogflow-webhook-example/main.py]: move to doc string

File: examples/dialogflow-webhook-example/main.py [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+                                     'user_id_action-followup')
+    parameter_full_name = '.'.join(['parameters', 'user_id'])
+    output_context_df = flattener.get_context()
+    user_id = output_context_df.loc[output_context_df.name == context_full_name,
+                                    parameter_full_name].to_list()[0]
+    return user_id
+
+
+class Intents:
+  """Defines intents."""
```
Comment (jaketf) [inline examples/dialogflow-webhook-example/main.py]: Mention that this is a model of dialog flow intents and add link to public docs.

**Comment by jaketf** [comment]
CI style check is failing please run `make fmt`.

**Comment by carolinahdez** [comment]
All the changes have been addressed

File: examples/dialogflow-webhook-example/main.py [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+                        ],
+                        'balance': 110
+                     }
+                 }
+             }
+        """
+            user_dict = user.to_dict()
+            message = 'From webhook: Here are all the transactions that I found:\n'
+            all_transactions = ','.join([
+                '{} {}'.format(transaction['type'], transaction['amount'])
```
Comment (jaketf) [inline examples/dialogflow-webhook-example/main.py:301]: Use f-string here and everywhere they are much faster, more concise.
```python3
f"{transaction['type']} {transaction['amount']}"
```
https://realpython.com/python-f-strings/

**Pull request merged**